tests/bunit.tests/Rendering/DeterministicRenderingTests.cs

```diff
@@ -0,0 +1,287 @@
+		var waitForTask2 = cut.WaitForAssertionAsync(() => cut.Instance.RendersCompleted.ShouldBe(3));
+
+		await cut.Instance.TriggerRender();
+		await waitForTask1;
+		await cut.Instance.TriggerRender();
+
+		await waitForTask2;
+	}
+
+	[Fact]
+	public async Task All_WaitFor_unsubscribe_blocks_dispatcher_queue()
+	{
+		var cut = Render<RemoteRenderTriggerComponent>();
+		var waitForTask1 = cut.WaitForAssertionAsync(() => cut.Instance.RendersCompleted.ShouldBe(2));
+		var waitForTask2 = cut.WaitForAssertionAsync(() => cut.Instance.RendersCompleted.ShouldBe(2));
+		await cut.Instance.TriggerRender();
+
+		var t1 = cut.InvokeAsync(() => throw new Exception("do nothing work item that emulates other work queued up"));
+
+		await waitForTask1;
+		await waitForTask2;
+		t1.Status.ShouldBe(TaskStatus.WaitingForActivation);
+		Renderer.DispatcherQueueLength.ShouldBe(1);
+	}
+
+	[Theory, AutoData]
+	public async Task WaitFor_breaks_when_queued_task_throws(Exception exceptedException)
+	{
+		var cut = Render<NoopComponent>();
+		_ = cut.InvokeAsync(() => throw exceptedException);
+
+		var waitFor = cut.WaitForStateAsync(() => cut.Instance.RendersCompleted == 2);
+
+		var actualException = await waitFor.ShouldThrowAsync<Exception>();
+		actualException.ShouldBe(exceptedException);
+	}
+
+	[Theory, AutoData]
+	public async Task WaitFor_breaks_when_scheduled_task_throws(Exception exceptedException)
+	{
+		var cut = Render<NoopComponent>();
+		var waitFor = cut.WaitForStateAsync(() => cut.Instance.RendersCompleted == 2);
+
+		_ = cut.InvokeAsync(() => throw exceptedException);
+
+		var actualException = await waitFor.ShouldThrowAsync<Exception>();
+		actualException.ShouldBe(exceptedException);
+	}
+
+	[Fact]
+	public async Task WaitFor_allows_each_queued_task_to_complete()
+	{
+		var cut = Render<RemoteRenderTriggerComponent>();
+		var completion = new TaskCompletionSource();
+		var t1 = cut.InvokeAsync(async () => { await completion.Task; await cut.Instance.TriggerRender(); });
+		var t2 = cut.InvokeAsync(() => throw new Exception("Should not be called"));
+
+		var waitFor = cut.WaitForStateAsync(() => cut.Instance.RendersCompleted == 2);
+		completion.SetResult();
+
+		await waitFor;
+		t2.Status.ShouldBe(TaskStatus.WaitingForActivation);
+	}
+
+	[Fact]
+	public void Disposing_Renderer_should_not_be_blocked()
+	{
+		var cut = Render<RemoteRenderTriggerComponent>();
+		var completion = new TaskCompletionSource();
+		var runningTask = cut.InvokeAsync(async () => { await completion.Task; await cut.Instance.TriggerRender(); });
+
+		DisposeComponents();
+
+		runningTask.Status.ShouldBe(TaskStatus.WaitingForActivation);
+	}
+
+	[Fact]
+	public async Task Triggering_renders_in_parent_is_not_influenced_by_child_component()
+	{
+		var cut = Render<RemoteRenderTriggerComponent>(p => p.AddChildContent<NeverFinishedComponent>());
+		var completion = new TaskCompletionSource();
+		var runningTask = cut.InvokeAsync(async () =>
+		{
+			await completion.Task;
```

Comment:
The reason why this test deadlocks is because the completion is never completed, i.e. via `completion.SetResult()`, so this is entirely reasonable.

Applying the suggestion below makes the test run to the end.

```suggestion
			// await completion.Task;
```

Reply:
Oh yes - shouldn't just copy&pated the test. Will update on push